kyo-core/shared/src/main/scala/kyo/typeMap.scala

```diff
@@ -10,7 +10,7 @@ extension [A](self: TypeMap[A])
         throw new RuntimeException(s"fatal: kyo.TypeMap of contents [$self] missing value of type: [${t.show}].")
 
     inline def get[B >: A](using inline t: Tag[B]): B =
-        self.getOrElse(t, fatal).asInstanceOf[B]
+        self.collectFirst { case (tag, value) if tag <:< t => value }.getOrElse(fatal).asInstanceOf[B]
```

Comment:
By doing this we don't have to explicitly widen subtypes. This is basically how ZEnvironment works.
However, since this type map is immutable and usually provided once at the _top of the application,_ ZEnvironment also keeps a cache of successful lookups, so only the first lookup needs to be O(n).

Reply:
Can you explain how the first lookup will be `O(1)`? I will cut a PR to add unit tests + attempt a slightly more optimized option.

Reply:
Oh, no, the first lookup needs to find the subtype by walking through each item (unless there's some other clever algorithm we can come up with). But once it finds the value, it'll put it in the cache.

Reply:
Maybe we could fist check if the tag isn't resolved directly from the map first and only check individual items if it returns false? It think it'd provide a fast path in many scenarios.

Reply:
Yep! That's why my PR aims to do https://github.com/getkyo/kyo/pull/427

Reply:
Cool! I hadn't noticed the initial lookup.